keps/sig-storage/1472-storage-capacity-tracking/README.md

```diff
@@ -281,63 +280,89 @@ still an unsolved problem.
-// driver, one topology segment, and the parameters of one storage class.
+// CSIStorageCapacity stores the result of one CSI GetCapacity call.
+// For a given StorageClass, this describes the available capacity in a
+// particular topology segment.  This can be used when considering where to
+// instantiate new PersistentVolumes.
+//
+// For example this can express things like:
+// - StorageClass "standard" has "1234 GiB" available in "topology.kubernetes.io/zone=us-east1"
+// - StorageClass "localssd" has "10 GiB" available in "kubernetes.io/hostname=knode-abc123"
+//
+// The following three cases all imply that no capacity is available for
+// a certain combination:
+// - no object exists with suitable topology and storage class name
+// - such an object exists, but the capacity is unset
+// - such an object exists, but the capacity is zero
+//
+// The producer of these objects can decide which approach is more suitable.
+//
+// They are consumed by the kube-scheduler when a CSI driver opts into capacity-aware
+// scheduling with CSIDriverSpec.StorageCapacity. The scheduler compares the
+// MaximumVolumeSize against the requested size of pending volumes to filter
+// out unsuitable nodes. If MaximumVolumeSize is unset, it falls back to
+// a comparison against the less precise Capacity. If that is also unset,
+// the scheduler assumes that capacity is insufficient and tries some other node.
 type CSIStorageCapacity struct {
-    metav1.TypeMeta
-    // Standard object's metadata. The name has no particular meaning and just has to
-    // meet the usual requirements (length, characters, unique). To ensure that
-    // there are no conflicts with other CSI drivers on the cluster, the recommendation
-    // is to use csisc-<uuid>.
-    //
-    // Objects are not namespaced.
-    //
-    // More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
-    // +optional
-    metav1.ObjectMeta
-
-    // Spec contains the fixed properties of one capacity value.
-    Spec CSIStorageCapacitySpec
-
-    // Status contains the properties that can change over time.
-    Status CSIStorageCapacityStatus
-}
-
-// CSIStorageCapacitySpec contains the fixed properties of one capacity value.
-// one capacity value.
-type CSIStorageCapacitySpec struct {
-    // The CSI driver that provides the storage.
-    // This must be the string returned by the CSI GetPluginName() call.
-    DriverName string
-
-    // NodeTopology defines which nodes have access to the storage for which
-    // capacity was reported. If not set, the storage is accessible from all
-    // nodes in the cluster.
-    // +optional
-    NodeTopology *v1.NodeSelector
-
-    // The storage class name of the StorageClass which provided the
-    // additional parameters for the GetCapacity call.
-    StorageClassName string
-}
-
-// CSIStorageCapacityStatus contains the properties that can change over time.
-type CSIStorageCapacityStatus struct {
```

Comment:
The YAML examples below still use this (old?) schema

Reply:
Good catch. I'll update those later.

For now let's see whether the autoscaler sections gets approval.